### PR TITLE
8279064: New options for ktab to provide non-default salt

### PR DESCRIPTION
**Pull request by wangweij**
Please review this enhancement and its [CSR](https://bugs.openjdk.java.net/browse/JDK-8279632). Two new options `-s salt` and `-f` can be specified on the `ktab` command when adding entries.

I'm a little concerned about the compatibility risk described in the CSR, i.e. the `-f` option is already used in `ktab -d` to force removing entries. Hopefully not many people are writing their own wrappers on ktab that always include the `-f` option. I do want to be consistent with the naming in the MIT krb5 ktutil command.

Another thing worth mentioning is the [KerberosKey:&lt;new&gt;(KerberosPrincipal principal, char[] password, String algorithm)](https://github.com/openjdk/jdk/blob/3790e58090be25421e3e323eb29deea100b7608c/src/java.security.jgss/share/classes/javax/security/auth/kerberos/KerberosKey.java#L149) constructor which always uses the default salt. For consistency, it looks like a new constructor should be added that takes the salt string as a parameter as well. However, I don't intend to add it as I cannot see a proper usage for it. In fact, I now regret adding the constructor linked above.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8279064](https://bugs.openjdk.java.net/browse/JDK-8279064): New options for ktab to provide non-default salt
 * [JDK-8279632](https://bugs.openjdk.java.net/browse/JDK-8279632): New options for ktab to provide non-default salt (**CSR**)

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6991/head:pull/6991` \
`$ git checkout pull/6991`

Update a local copy of the PR: \
`$ git checkout pull/6991` \
`$ git pull https://git.openjdk.java.net/jdk pull/6991/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6991`

View PR using the GUI difftool: \
`$ git pr show -t 6991`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6991.diff">https://git.openjdk.java.net/jdk/pull/6991.diff</a>

</details>

### CODE REVIEW

**Comment by valeriepeng** [comment]
I will take a look. Thanks~

**Comment by valeriepeng** [comment]
I added myself to the CSR as reviewer. Made minor edits, re-ordered the command line option ordering based on the ordering mentioned in the description.

File: src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java [hunk 1]
```diff
@@ -434,11 +454,13 @@ void printHelp() {
         System.out.println();
         System.out.println("-l [-e] [-t]\n"
                 + "    list the keytab name and entries. -e with etype, -t with timestamp.");
-        System.out.println("-a <principal name> [<password>] [-n <kvno>] [-append]\n"
+        System.out.println("-a <principal name> [<password>] [-s <salt>] [-n <kvno>] [-f | -s <salt>] [-append]\n"
```
Comment (valeriepeng) [inline src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java:457]: The first "[-s <salt>]" looks redundant? Hmm, I edited your CSR from [-s <salt>] to [-s <salt> | -f]. Maybe you want to change that to be consistent with what you have here.

File: src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (valeriepeng) [inline src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java:2]: 2022

Comment (valeriepeng) [reply src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java:2]: Same goes for sun/security/krb5/internal/ktab/KeyTab.java.

File: test/jdk/sun/security/krb5/auto/tools/KtabSalt.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/krb5/auto/tools/KtabSalt.java:2]: Same goes for test/jdk/sun/security/krb5/auto/Context.java.

[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/krb5/auto/tools/KtabSalt.java:2]: And test/jdk/sun/security/krb5/tools/KtabCheck.java.

File: src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java (hunk 1 above)
Comment (wangweij) [reply src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java:457]: Oops, no idea why I typed twice. Thanks for fixing the CSR.

File: src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java (hunk 2 above)
Comment (wangweij) [reply src/java.security.jgss/windows/classes/sun/security/krb5/internal/tools/Ktab.java:2]: Yes.

File: test/jdk/sun/security/krb5/auto/tools/KtabSalt.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangweij) [reply test/jdk/sun/security/krb5/auto/tools/KtabSalt.java:2]: Yes.

**Review by valeriepeng — APPROVED** [review]
Looks good. Thanks~

**Comment by wangweij** [comment]
I heard the CSR is approved.

**Pull request closed without merging**